sdk-extension/opentelemetry-sdk-extension-aws/src/opentelemetry/sdk/extension/aws/trace/propagation/aws_xray_format.py

```
@@ -0,0 +1,283 @@
+    TRACE_ID_DELIMITER = "-"
+    TRACE_ID_DELIMITER_INDEX_1 = 1
+    TRACE_ID_DELIMITER_INDEX_2 = 10
+    TRACE_ID_FIRST_PART_LENGTH = 8
+
+    PARENT_ID_KEY = "Parent"
+    PARENT_ID_LENGTH = 16
+
+    SAMPLED_FLAG_KEY = "Sampled"
+    SAMPLED_FLAG_LENGTH = 1
+    IS_SAMPLED = "1"
+    NOT_SAMPLED = "0"
+
+    def extract(
+        self,
+        getter: Getter[TextMapPropagatorT],
+        carrier: TextMapPropagatorT,
+        context: typing.Optional[Context] = None,
+    ) -> Context:
+        trace_header_list = getter(carrier, self.TRACE_HEADER_KEY)
+
+        if not trace_header_list or len(trace_header_list) != 1:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        trace_header = trace_header_list[0]
+
+        if not trace_header:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        trace_id, span_id, sampled, err = self.extract_span_properties(
+            trace_header
+        )
+
+        if err is not None:
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        options = 0
+        if sampled:
+            options |= trace.TraceFlags.SAMPLED
+
+        span_context = trace.SpanContext(
+            trace_id=trace_id,
+            span_id=span_id,
+            is_remote=True,
+            trace_flags=trace.TraceFlags(options),
+            trace_state=trace.TraceState(),
+        )
+
+        if not span_context.is_valid:
+            _logger.error(
+                "Invalid Span Extracted. Insertting INVALID span into provided context."
+            )
+            return trace.set_span_in_context(
+                trace.INVALID_SPAN, context=context
+            )
+
+        return trace.set_span_in_context(
+            trace.DefaultSpan(span_context), context=context
+        )
+
+    def extract_span_properties(self, trace_header):
```

another way you could write this function: raise a custom exception for any invalid parsing cases, and log and set the context to invalid in the caller. 

I see a lot of complexity in the multiple break statements which are intended to stop execution of parsing on first error encountered. This is pretty close to either raising an exception, or returning an error value.

Since you can't return just an error value, I think an exception would work.

---

I think that's a great idea! I felt a bit wary about it because it required creating such a simple custom exception, but I also think it makes a lot of sense in this case. Thanks for another great suggestion :)